contracts/src/v0.8/upkeeps/GovernanceUpkeep.sol

```diff
@@ -0,0 +1,137 @@
+      } else if (
+        s_governanceTokenContract.getPriorVotes(proposer, sub256(block.number, 1)) <
+        s_governanceContract.proposalThreshold()
+      ) {
+        performData = abi.encode(Action.CANCEL, i);
+        return (true, performData);
+      }
+    }
+
+    return (false, "");
+  }
+
+  ///@notice Chainlink Keepers will execute when checkUpkeep returns 'true'. Decodes the 'performData' passed in from checkUpkeep and performs an action as needed.
+  ///@param performData bytes encoded: (governance action required, index of proposal)
+  ///@dev The governance contract has action validation built-in
+  function performUpkeep(bytes calldata performData) external override {
+    //Decode performData
+    (Action performAction, uint256 proposalIndex) = abi.decode(performData, (Action, uint256));
+
+    //Check state of proposal at index
+    GovernorAlpha.ProposalState state = s_governanceContract.state(proposalIndex);
+
+    //Revalidate state and action of provided index
+    if (performAction == Action.QUEUE && state == GovernorAlpha.ProposalState.Succeeded) {
+      s_governanceContract.queue(proposalIndex);
+    } else if (performAction == Action.EXECUTE && state == GovernorAlpha.ProposalState.Queued) {
+      s_governanceContract.execute(proposalIndex);
+    } else if (performAction == Action.CANCEL && state != GovernorAlpha.ProposalState.Executed) {
+      s_governanceContract.cancel(proposalIndex);
+    } else if (performAction == Action.UPDATE_INDEX) {
+      uint256 newStartingIndex = findStartingIndex();
+      require(newStartingIndex > s_proposalStartingIndex, "No update required");
+      s_proposalStartingIndex = newStartingIndex;
+    }
+  }
+
+  ///@notice Goes through each proposal, if any proposal is in a state that needs to be checked OR its the last proposal, will set it as the starting proposal index for future checks
+  ///@return index The proposal index to start checking from
+
+  function findStartingIndex() public view returns (uint256 index) {
+    // Set current starting index
+    uint256 pendings_proposalStartIndex = s_proposalStartingIndex;
+    // Get current proposal count
+    uint256 proposalCount = s_governanceContract.proposalCount();
+
+    for (uint256 i = pendings_proposalStartIndex; i <= proposalCount; i++) {
+      GovernorAlpha.ProposalState state = s_governanceContract.state(i);
+      if (
+        state == GovernorAlpha.ProposalState.Pending ||
+        state == GovernorAlpha.ProposalState.Active ||
+        state == GovernorAlpha.ProposalState.Succeeded ||
+        state == GovernorAlpha.ProposalState.Queued ||
+        i == proposalCount
+      ) {
+        pendings_proposalStartIndex = i;
+        break;
+      }
+    }
+    return (pendings_proposalStartIndex);
+  }
+
+  function sub256(uint256 a, uint256 b) internal pure returns (uint256) {
+    require(b <= a, "subtraction underflow");
+    return a - b;
+  }
+}
+
+interface GovernorAlphaToken {
```

Comment:
nitpick, but I'd usually expect to see this at the top of the file, before the contract definition starts